accepted/wasmtime-baseline-compilation.md

```diff
@@ -0,0 +1,259 @@
+# Baseline Compilation in Wasmtime
+
+Authors: Saúl Cabrera (@saulecabrera); Chris Fallin (@cfallin)
+
+## Summary
+
+This RFC proposes the addition of a new WebAssembly (Wasm) compiler to Wasmtime:
+a single-pass or “baseline” compiler. A baseline compiler works to improve
+overall compilation performance, yielding faster startup times, at the cost of
+less optimized code.
+
+
+## Motivation
+
+Wasmtime currently uses Cranelift by default, which is an optimizing compiler.
+Cranelift performs code optimizations at the expense of slower compilation
+times. This makes Just-In-Time (JIT) compilation of Wasm unsuitable for cases
+where higher compilation performance is desired (e.g. short-lived trivial
+programs, cases in which startup time is more critical than runtime
+performance). 
+
+The introduction of a baseline compiler is a first step towards: (i) faster
+compilation and startup times (ii) enabling a tiered compilation model in
+Wasmtime, similar to what is present in Wasm engines in web browsers. This RFC
+**does not** account for tiered compilation, it only accounts for the
+introduction of a baseline compiler.
+
+**Approximate** [measurements](https://github.com/Shopify/wasm-bench) taken on
+top of a subset of the [Sightglass benchmark
+suite](https://github.com/bytecodealliance/sightglass/tree/main/benchmarks)
+– using different optimizing and baseline compilers (Cranelift from Wasmtime,
+Liftoff from V8 and RabaldrMonkey from SpiderMonkey) – show that a baseline
+compiler on average yields 15x to 20x faster compilation while producing code
+that is on average 1.1x to 1.5x slower than the one produced by an optimizing
+compiler. These measurements align on average with other measurements observed
+when comparing interpretation and compilation for WebAssembly[^1].
+
+
+[^1]: Ben L. Titzer. [A fast in-place interpreter for
+  WebAssembly](https://arxiv.org/pdf/2205.01183.pdf)
+
+## Proposal: Winch, a baseline compiler for Wasmtime
+
+Winch: WebAssembly Intentionally-Non-Optimizing Compiler and Host
+
+
+### Design Principles
+
+* Single pass over Wasm bytecode
+* Function as the unit of compilation
+* Machine code generation directly from Wasm bytecode – no intermediate
+  representation
+* Avoid reinventing machine-code emission – use Cranelift's instruction emitter
+  code to create an assembler library ("MacroAssembler")
+* Prioritize compilation performance over runtime performance
+
+### High-level overview
+
+```mermaid
+  graph TD;
+  A(wasmparser)-->B(cranelift-wasm);
+  A-->C(winch);
+  C-->D(MacroAssembler);
+  D-->X(cranelift-asm);
+  X-->E(MachInst);
+  X-->F(MachBuffer);
+  B-->G(cranelift);
+  G-->X;
+```
+
+### MacroAssembler and Borrowing from Cranelift
+
+We plan to factor out the lower layers of Cranelift that produce and operate on
+machine code in order to reuse them as a generic assembler library
+(“MacroAssembler”).
```

Comment:
This is really mostly a nitpick from myself, but I think we're confusing two different levels of abstraction here. Provided the examples below, namely the higher-level API call, I think this is really an `Assembler` library. From my understanding of Spidermonkey's code base, a `MacroAssembler` tries to hide some extra complexity by only having "operands" (and details about memory addresses/registers are hidden by the abstraction). Or imbuing some other specific semantics of operators that don't exist at the assembler layer; think for instance about having some `wasm_div` method, that would trap if the divisor is zero; there's no assembly instructions to do so, but a macro-assembler would combine different assembler method calls to generate the code that implements this behavior. As a matter of fact, the macro-assembler is one layer of abstraction above an assembler, and what's described here rather looks like a plain assembler to me, unless I'm mistaken.

Reply:
> From my understanding of Spidermonkey's code base, a MacroAssembler tries to hide some extra complexity by only having "operands" (and details about memory addresses/registers are hidden by the abstraction).

My understanding is a bit different here. As far as I can tell, the MacroAssembler aside from operating on[ top of the concept of operands](https://searchfox.org/mozilla-central/source/js/src/jit/MacroAssembler.h#530), also operates on top of other abstractions like [Register](https://searchfox.org/mozilla-central/source/js/src/jit/MacroAssembler.h#1046), [Imm, and Address](https://searchfox.org/mozilla-central/source/js/src/jit/MacroAssembler.h#1048); still encapsulating most of the architecture-specific details.

As far as encapsulating complexity, my mental model doesn't diverge from what you've outlined: emitting a set of assembly assembly instructions representing a given wasm instruction, with the difference that we're not strictly relying on/introducing  the presence of an Assembler for lowering; but instead potentially relying on helper methods.

From a strict point of view, at least taking in to account SpiderMonkey's implementation, maybe MacroAssembler is not the right term, but eventually it could grow to be the right term. For the initial implementation introducing an Assembler abstraction seems a bit premature IMO. That said, I wouldn't be opposed to naming this abstraction `Assembler` instead. 

@cfallin do you have any further thoughts that you'd like to share according to your experience with SpiderMonkey?

Reply:
Yes, I agree generally with Ben here that what we want initially is not-so-macro; actually I think as a first approximation, one method call per `MachInst` is about right. At that level, we'll need different wasm-opcode-to-assembler-API code for each architecture. Details like addressing modes, and two-register (x86) vs three-register (aarch64) form, are best dealt with explicitly at this level, rather than being papered over; for example an alternative approach to the 2/3-register-form difference would be to emit a mov/add sequence for every add and emulate 3-reg form on x86, but that's a significant pessimization.

But I also think there's a lot of room to evolve a general set of abstractions on top of this and share implementations where necessary, especially for the more complex lowerings like (for example) reference-type read/write barrier fastpaths, and other ops built from primitives. We could implement something like a `CommonAssembler` trait that provides load/store/move/add/sub/compare/branch/... and then write another layer that is generic over that.

I think that starting point and evolutionary path is the important point and the names aren't so important; so `Assembler` rather than `MacroAssembler` is fine, if that's the way others prefer!